src/lib/y2firewall/ui_state.rb

```diff
@@ -0,0 +1,155 @@
+    #
+    # It remembers the decision so the user is taken back to a sensible point of
+    # the tree (very often the last he decided to visit) after redrawing.
+    #
+    # @param [CWM::Page] page associated to the tree node
+    def go_to_tree_node(page)
+      self.candidate_nodes =
+        if page.respond_to?(:zone)
+          zone_page_candidates(page)
+        else
+          [page.label]
+        end
+      # Landing in a new node, so invalidate previous details about position
+      # within a node, they no longer apply
+      self.tab = nil
+    end
+
+    # Method to be called when the user switches to a tab within a tree node.
+    #
+    # It remembers the decision so the same tab is showed in case the user stays
+    # in the same node after redrawing.
+    #
+    # @param [CWM::Page] page associated to the tab
+    def switch_to_tab(page)
+      self.tab = page.label
+    end
+
+    # Method to be called when the user operates in a row of a table of zones
+    # or creates a new zone.
+    #
+    # @param zone [String] zone name
+    def select_row(zone_name)
+      self.row_id = zone_name
+    end
+
+    # Select the page to open in the general tree after a redraw
+    #
+    # @param pages [Array<CWM::Page>] all the pages in the tree
+    # @return [CWM::Page, nil]
+    def find_tree_node(pages)
+      candidate_nodes.each.with_index do |candidate, idx|
+        result = pages.find { |page| matches?(page, candidate) }
+        if result
+          # If we had to use one of the fallbacks, the tab name is not longer
+          # trustworthy
+          self.tab = nil unless idx.zero?
+          return result
+        end
+      end
+      self.tab = nil
+      nil
+    end
+
+    # Select the tab to open within the node after a redraw
+    #
+    # @param pages [Array<CWM::Page>] pages for all the possible tabs
+    # @return [CWM::Page, nil]
+    def find_tab(pages)
+      return nil unless tab
+      pages.find { |page| page.label == tab }
+    end
+
+    # @see #row_id
+    attr_accessor :row_id
+
+  protected
+
+    # Where to place the user within the general tree in next redraw
+    # @return [Array<Integer, String>]
+    attr_accessor :candidate_nodes
+
+    # Concrete tab within the current node to show in the next redraw
+    # @return [String, nil]
+    attr_reader :tab
+    # @see #tab
+    def tab=(tab)
+      @tab = tab
+      # If the user switched to a new tab, invalidate details about the inner
+      # table
+      self.row_id = nil
+    end
+
+    # List of candidate nodes to go back after opening a zone view in the tree
+    def zone_page_candidates(page)
+      zone = page.zone
+
+      [zone.name]
+    end
+
+    # Whether the given page matches with the candidate tree node
+    #
+    # @param page [CWM::Page]
+    # @param candidate [Integer, String]
+    # @return boolean
+    def matches?(page, candidate)
+      page.label == candidate
+    end
+
+    class << self
+      # Singleton instance
```

Comment:
Why not use the stdlib https://devdocs.io/ruby~2.5/singleton ?

Reply:
I do not know, this code comes from yast2-storage-ng. But I do not see a good reason.